src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Options/Option2`1.cs

```diff
@@ -0,0 +1,135 @@
+        /// </summary>
+        public string Name { get; }
+
+        /// <summary>
+        /// The default value of the option.
+        /// </summary>
+        public T DefaultValue { get; }
+
+        /// <summary>
+        /// The type of the option value.
+        /// </summary>
+        public Type Type => typeof(T);
+
+        public ImmutableArray<OptionStorageLocation2> StorageLocations { get; }
+
+        [Obsolete("Use a constructor that specifies an explicit default value.")]
+        public Option2(string feature, string name)
+            : this(feature, name, default!)
+        {
+            // This constructor forwards to the next one; it exists to maintain source-level compatibility with older callers.
+        }
+
+        public Option2(string feature, string name, T defaultValue)
+            : this(feature, name, defaultValue, storageLocations: Array.Empty<OptionStorageLocation2>())
+        {
+        }
+
+        public Option2(string feature, string name, T defaultValue, params OptionStorageLocation2[] storageLocations)
+            : this(feature, group: OptionGroup.Default, name, defaultValue, storageLocations)
+        {
+        }
+
+        internal Option2(string feature, OptionGroup group, string name, T defaultValue, params OptionStorageLocation2[] storageLocations)
+            : this(feature, group, name, defaultValue, storageLocations.ToImmutableArray())
+        {
+        }
+
+        internal Option2(string feature, OptionGroup group, string name, T defaultValue, ImmutableArray<OptionStorageLocation2> storageLocations)
+        {
+            if (string.IsNullOrWhiteSpace(feature))
+            {
+                throw new ArgumentNullException(nameof(feature));
+            }
+
+            if (string.IsNullOrWhiteSpace(name))
+            {
+                throw new ArgumentException(nameof(name));
+            }
+
+            this.Feature = feature;
+            this.Group = group ?? throw new ArgumentNullException(nameof(group));
+            this.Name = name;
+            this.DefaultValue = defaultValue;
+            this.StorageLocations = storageLocations;
+        }
+
+#if CODE_STYLE
+        object? IOption2.DefaultValue => this.DefaultValue;
+
+        bool IOption2.IsPerLanguage => false;
+#else
+        object? IOption.DefaultValue => this.DefaultValue;
+
+        bool IOption.IsPerLanguage => false;
+
+        ImmutableArray<OptionStorageLocation> IOption.StorageLocations
+            => this.StorageLocations.As<OptionStorageLocation>();
+#endif
+
+        OptionGroup IOptionWithGroup.Group => this.Group;
+
+        public override string ToString()
+        {
+            return string.Format("{0} - {1}", this.Feature, this.Name);
+        }
+
+        public static implicit operator OptionKey2(Option2<T> option)
+        {
+            return new OptionKey2(option);
+        }
+
+#if !CODE_STYLE
+        public static implicit operator Option<T>(Option2<T> option)
```

Comment:
💡 I would prefer to see an internal extension method in the workspaces layer called where necessary, instead of an implicit operator with multiple allocations for each usage.